### PR TITLE
feat(json-schema): Add discriminator support and oneOf/anyOf union option

### PR DESCRIPTION
**Pull request by mike-taylor99**
## Summary
Adds support for discriminator properties in JSON Schema emitter with an opt-in oneOf/anyOf union generation option.

## Changes
- Automatically injects discriminator property into base models with `@discriminator` decorator
- Marks discriminator property as required in generated schemas
- Adds new `polymorphic-models-strategy` emitter option to generate oneOf or anyOf schemas for closed discriminated unions
- Includes discriminator.mapping for improved validation

## Testing
- Added 11 comprehensive tests covering:
  - Basic discriminator property injection
  - Required field marking
  - OneOf union generation with mapping
  - Edge cases and validation

## Documentation
All tests passing. Changeset included for release notes.

### CODE REVIEW

File: packages/json-schema/test/discriminator.test.ts [hunk 1]
```diff
@@ -0,0 +1,355 @@
+import { deepStrictEqual, ok, strictEqual } from "assert";
+import { describe, it } from "vitest";
+import { emitSchema } from "./utils.js";
+
+describe("json-schema: discriminator", () => {
```
Comment (timotheeguerin) [inline packages/json-schema/test/discriminator.test.ts]: you can remove this top level describe, we used to have that before when mocha would merge all the test together but now its already scopped per file

File: packages/json-schema/src/lib.ts [hunk 2]
```diff
@@ -62,6 +62,13 @@ export interface JSONSchemaEmitterOptions {
    * @defaultValue false
    */
   "seal-object-schemas"?: boolean;
+
+  /**
+   * When true, models with the discriminator decorator that have derived models will emit a oneOf schema
+   * with references to all derived models, creating a closed discriminated union.
+   * @defaultValue false
+   */
+  "emit-discriminated-union"?: boolean;
```
Comment (timotheeguerin) [inline packages/json-schema/src/lib.ts]:
```suggestion
  "polymorphic-models-strategy"?: "ignore" | "oneOf" | "anyOf";
```

makes it clearer and allow for specifying if using `anyOf` instead too which I know some tools prefer

File: .chronus/changes/main-2025-10-19-17-48-27.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+changeKind: feature
+packages:
+  - "@typespec/json-schema"
+---
+
+add discriminator support and oneOf union option
```
Comment (timotheeguerin) [inline .chronus/changes/main-2025-10-19-17-48-27.md]:
```suggestion
Add support for emitting polymorhic models as `oneOf`
```

+ "or `anyOf`" if doing change below

File: packages/json-schema/test/discriminator.test.ts [hunk 4]
```diff
@@ -0,0 +1,394 @@
+import { deepStrictEqual, ok, strictEqual } from "assert";
+import { describe, it } from "vitest";
+import { emitSchema } from "./utils.js";
+
+describe("model inheritance with @discriminator", () => {
```
Comment (timotheeguerin) [inline packages/json-schema/test/discriminator.test.ts]:
```suggestion
describe("default behavior", () => {
```
does this make it clearer?

File: packages/json-schema/src/json-schema-emitter.ts [hunk 5]
```diff
@@ -653,6 +680,43 @@ export class JsonSchemaEmitter extends TypeEmitter<Record<string, any>, JSONSche
...
+      const prop = derived.properties.get(discriminator.propertyName);
+      if (prop?.type.kind === "String") {
+        const value = (prop.type as StringLiteral).value;
+        mapping[value] = `${derived.name}.json`;
+      }
+    }
+
+    const schema = this.#initializeSchema(model, name, {
+      [strategy]: variants,
+      discriminator: {
```
Comment (timotheeguerin) [inline packages/json-schema/src/json-schema-emitter.ts]: missed that the first time but `discriminator` is not part of json schema is it?

File: packages/json-schema/src/json-schema-emitter.ts [hunk 6]
```diff
@@ -653,6 +680,43 @@ export class JsonSchemaEmitter extends TypeEmitter<Record<string, any>, JSONSche
...
+
+      // Add reference to each derived model
+      const derivedRef = this.emitter.emitTypeReference(derived);
+      variants.push(derivedRef);
+
+      // Extract discriminator value from derived model
+      const prop = derived.properties.get(discriminator.propertyName);
+      if (prop?.type.kind === "String") {
+        const value = (prop.type as StringLiteral).value;
+        mapping[value] = `${derived.name}.json`;
```
Comment (timotheeguerin) [inline packages/json-schema/src/json-schema-emitter.ts]: I think this won't work if the schema is not in its own file(e.g. using bundleId)

File: packages/json-schema/src/json-schema-emitter.ts (hunk 5 above)
Comment (mike-taylor99) [reply packages/json-schema/src/json-schema-emitter.ts]: you're right, we just need the `oneOf` / `anyOf` support

File: packages/json-schema/test/discriminator.test.ts
Comment (timotheeguerin) [inline packages/json-schema/test/discriminator.test.ts:1]: While testing noticed an issue, the properties on the base class are dropped

[pr playground](https://cadlplayground.z22.web.core.windows.net/prs/9038/?e=%40typespec%2Fjson-schema&options=%7B%22options%22%3A%7B%22%40typespec%2Fopenapi3%22%3A%7B%22file-type%22%3A%22yaml%22%7D%2C%22%40typespec%2Fjson-schema%22%3A%7B%22polymorphic-models-strategy%22%3A%22oneOf%22%7D%7D%7D&vs=%7B%7D&c=aW1wb3J0ICJAdHlwZXNwZWMvanNvbi1zY2hlbWEiOwoKQEpzb25TxQ8uxBvGCwpuYW1lc3BhY2UgVGVzdMQoZGlzY3JpbWluYXRvcigia2luZCIpCm1vZGVsIFBldCB7CiAgY29tbW9uOiBzdHJpbmc7Cn0KxyFDYXQgZXh0ZW5kc8ktxEI6ICJjYXQixSrQYmRvZ0vMZURvZ9hEZG9nIjsKICBiYXJrOiBib29sZWFuy39CdWxsZMs%2FxEvEP8djOiAiYnVsbERvZyIKfQ%3D%3D)
 ```tsp
@discriminator("kind")
model Pet {
  common: string;
}

model Cat extends Pet {
  kind: "cat";
}
```

`common` is not on `Pet` anymore
